File: resto-pulse-mvp/backend/app/services/turkish_text_fold.py
"""Turkce metin — sesli/yazili komut eslestirme icin ASCII yakinlastirma."""

from __future__ import annotations

import re

_NON_WORD = re.compile(r"[^\w\s]", re.UNICODE)


def normalize_tr_text(value: str) -> str:
    text = (value or "").strip().casefold()
    text = text.replace("’", "").replace("'", "")
    text = _NON_WORD.sub(" ", text)
    return " ".join(text.split())


def fold_tr_ascii(value: str) -> str:
    text = normalize_tr_text(value)
    return (
        text.replace("ı", "i")
        .replace("ğ", "g")
        .replace("ü", "u")
        .replace("ş", "s")
        .replace("ö", "o")
        .replace("ç", "c")
    )


def text_includes_tr_folded(haystack: str, needle: str, *, min_needle_len: int = 3) -> bool:
    folded_needle = fold_tr_ascii(needle)
    if len(folded_needle) < min_needle_len:
        return False
    return folded_needle in fold_tr_ascii(haystack)

File: resto-pulse-mvp/mobile/lib/turkish-text-fold.ts
/** Turkce metin — konusma/yazim farklari icin ASCII yakinlastirma. */

export function normalizeTrSpeechText(value: string): string {
  const input = typeof value === 'string' ? value : '';
  return input
    .toLocaleLowerCase('tr-TR')
    .replace(/[’']/g, '')
    .replace(/[^\p{L}\p{N}\s.,]/gu, ' ')
    .replace(/\s+/g, ' ')
    .trim();
}

/** sütlaç ~ sutlac, cantık ~ cantik, şişe ~ sise */
export function foldTrAscii(value: string): string {
  return normalizeTrSpeechText(value)
    .replace(/ı/g, 'i')
    .replace(/ğ/g, 'g')
    .replace(/ü/g, 'u')
    .replace(/ş/g, 's')
    .replace(/ö/g, 'o')
    .replace(/ç/g, 'c');
}

export function textIncludesTrFolded(haystack: string, needle: string, minNeedleLen = 3): boolean {
  const foldedHaystack = foldTrAscii(haystack);
  const foldedNeedle = foldTrAscii(needle);
  if (!foldedNeedle || foldedNeedle.length < minNeedleLen) return false;
  return foldedHaystack.includes(foldedNeedle);
}

/**
 * iOS TTS ozel harfleri (o, u, s, …) tek tek okuyabiliyor.
 * Konusma icin ASCII yakinlastirma — ekranda orijinal metin kalir.
 */
export function prepareTurkishSpeechText(value: string): string {
  const input = typeof value === 'string' ? value : '';
  return foldTrAscii(input)
    .replace(/[&/\\|+_]/g, ' ')
    .replace(/\s+/g, ' ')
    .trim();
}
